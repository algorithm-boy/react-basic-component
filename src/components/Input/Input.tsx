import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { listenKeyboard, clearKeyboardListener } from 'peeler-js';
import { ListenerCallBack } from 'peeler-js/dist/listenKeyboard/index.d';
import { Style, GeneralProps } from '../global.d';
import { createPropsGetter, logger } from '../../utils';

export type autoFocusFn = (handleFocus: (isFocus: boolean) => void) => any;

export interface InputProps extends GeneralProps {
  type?: 'text' | 'password';
  id?: string;
  prefixCls?: string;
  placeholder?: string;
  value?: string;
  defaultValue?: string;
  maxLength?: number;
  clipboardFree?: boolean;
  copyFree? : boolean;
  pasteFree? : boolean;
  cutFree? : boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>, value: string) => any;
  onFocus?: (e: React.ChangeEvent<HTMLInputElement>) => any;
  onBlur?: (e: React.ChangeEvent<HTMLInputElement>) => any;
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => any;
  onPressEnter?: (e: React.KeyboardEvent<HTMLInputElement>) => any;
  onPaste?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  onContextMenu?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  onCopy?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  onCut?: (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>) => any;
  onRise?: ListenerCallBack;
  onFold?: ListenerCallBack;
  autoFocus?: autoFocusFn | boolean;
};

export interface InputStates {
  value?: string;
};

const defaultProps = {
  style: {} as any as Style,
  className: '',
  type: 'text',
  prefixCls: 'hfdc-input',
  maxLength: Infinity,
  clipboardFree: true,
  copyFree: true,
  pasteFree: true,
  cutFree: true,
  autoFocus: false as autoFocusFn | boolean
};

type DefaultProps = Readonly<typeof defaultProps>;

const getProps = createPropsGetter(defaultProps);

class Input extends PureComponent<InputProps & DefaultProps, InputStates> {
  static defaultProps = defaultProps;
  protected nodeRef: React.RefObject<HTMLInputElement>;

  constructor (props: InputProps & DefaultProps) {
    super(props);
    this.state = {
      value: (typeof props.value === 'undefined' ? props.defaultValue : props.value) || ''
    };

    this.nodeRef = React.createRef<HTMLInputElement>();
    this.handleChange = this.handleChange.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleClipboard = this.handleClipboard.bind(this);
    this.manualFocus = this.manualFocus.bind(this);

    if (typeof props.autoFocus === 'function') {
      props.autoFocus(this.manualFocus);
    }
  }

  static getDerivedStateFromProps(nextProps: InputProps) {
    if ('value' in nextProps) {
      return {
        value: nextProps.value
      };
    }
    return null;
  }

  public componentDidMount () {
    const node = this.nodeRef.current as HTMLInputElement;
    const { onRise, onFold, autoFocus } = getProps(this.props);

    if (autoFocus && typeof autoFocus === 'boolean') {
      // auto focus when trigger componentDidMount life-circle
      this.manualFocus(autoFocus);
    }

    listenKeyboard(node, onRise, onFold);
  }

  public componentWillUnmount () {
    const node = this.nodeRef.current as HTMLInputElement;
    clearKeyboardListener(node);
  }

  private handleChange (e: React.ChangeEvent<HTMLInputElement>) {
    const { onChange } = getProps(this.props);
    const currentVal = e.target.value;

    if (!('value' in this.props)) {
      this.setState({ value: currentVal });
    }

    onChange && onChange(e, currentVal);
  }

  private handleFocus (e: React.FocusEvent<HTMLInputElement>) {
    const { onFocus } = getProps(this.props);
    onFocus && onFocus(e);
  }

  public manualFocus (isFocus: boolean): void {
    const node = this.nodeRef.current;
    if (!node) return logger.logWarn('Input - cannot evoke the function because the DOM not initialize yet');
    if (!!isFocus) {
      node.focus();
    } else {
      node.blur();
    }
  }

  private handleBlur (e: React.FocusEvent<HTMLInputElement>) {
    const { onBlur } = getProps(this.props);
    onBlur && onBlur(e);
  }

  private handleKeyDown (e: React.KeyboardEvent<HTMLInputElement>): void {
    const { onPressEnter, onKeyDown } = this.props;
  
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  }

  private handleClipboard (e: React.MouseEvent<HTMLInputElement, MouseEvent> | React.ClipboardEvent<HTMLInputElement>): void {
    const { onPaste, onContextMenu, onCopy, onCut, clipboardFree, copyFree, pasteFree, cutFree } = this.props;

    switch (('' + e.type).toLowerCase()) {
      case 'copy':
        onCopy && onCopy(e);
        !copyFree && notFree();
        break;
      case 'paste':
        onPaste && onPaste(e);
        !pasteFree && notFree();
        break;
      case 'cut':
        onCut && onCut(e);
        !cutFree && notFree();
        break;
      case 'contextmenu':
        onContextMenu && onContextMenu(e);
        break;
    }

    !clipboardFree && notFree();

    function notFree () {
      if (e && e.preventDefault) {
        e.preventDefault();
      } else if (typeof window !== 'undefined') {
        (window as Window).event!.returnValue = false;
      }
    }
  }

  public render() {
    const {
      type,
      id,
      maxLength,
      className,
      prefixCls,
      style,
      placeholder
    } = getProps(this.props);

    const { value } = this.state;

    const classes = classnames({
      [prefixCls]: true,
      [className || '']: !!className
    });

    return (
      <input
        type={type}
        id={id}
        style={style}
        value={value}
        maxLength={maxLength}
        className={classes}
        placeholder={placeholder}
        onChange={this.handleChange}
        onFocus={this.handleFocus}
        onBlur={this.handleBlur}
        onKeyDown={this.handleKeyDown}
        onPaste={this.handleClipboard}
        onContextMenu={this.handleClipboard}
        onCopy={this.handleClipboard}
        onCut={this.handleClipboard}
        ref={this.nodeRef}
      />
    );
  };
};

export default Input;
