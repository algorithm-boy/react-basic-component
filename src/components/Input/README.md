# Input

## API

| 参数(props) | 说明(specification) | 类型(type) | 默认值(defaultProps) |
| --- | --- | --- | --- |
| className(optional) | Input的类名 (The class property for Input) | string | - |
| style(optional) | Input的行内样式 (The inline style for Input) | object | {} |
| type(optional) | Input的type属性 (The type property for input) | string | text |
| id(optional) | ID (The ID for input) | string | - |
| prefixCls(optional) | 类名前缀 (The prefix of class) | string | 'hfdc-input' |
| placeholder(optional) | 占位符 (The placeholder) | string | - |
| value(optional) | 文本域的值 (The input content value) | string | - |
| defaultValue(optional) | 文本域的初始值，优先级低于 `value` (The initial input content value, priority is lower than `value`) | string | - |
| maxLength(optional) | 文本长度限制 (The max limit of content length) | number | Infinity |
| clipboardFree(optional) | 是否启用剪贴板 (The input whether or not enable clipboard) | boolean | true |
| copyFree(optional) | 是否启用复制 (The input whether or not enable copy) | boolean | true |
| pasteFree(optional) | 是否启用粘贴 (The input whether or not enable paste) | boolean | true |
| cutFree(optional) | 是否启用剪切 (The input whether or not enable cut) | boolean | true |
| onChange(optional) | 文本域发生变化时的回调 (Trigger when input content changed) | Function(e: React.ChangeEvent, value: string) | - |
| onFocus(optional) | 聚焦的回调 (Trigger when focus) | Function(e: React.ChangeEvent) | - |
| onBlur(optional) | 失焦的回调 (Trigger when blur) | Function(e: React.ChangeEvent) | - |
| onKeyDown(optional) | 按键的回调 (Trigger when keydown) | Function(e: React.KeyboardEvent) | - |
| onPressEnter(optional) | 按回车键的回调 (Trigger when Enter press down) | Function(e: React.KeyboardEvent) | - |
| onPaste(optional) | 粘贴的回调 (Trigger when paste) | Function(e: React.MouseEvent \| React.ClipboardEvent) | - |
| onContextMenu(optional) | 打开上下文菜单的回调 (Trigger when open the ContextMenu) | Function(e: React.MouseEvent \| React.ClipboardEvent) | - |
| onCopy(optional) | 复制的回调 (Trigger when copy) | Function(e: React.MouseEvent \| React.ClipboardEvent) | - |
| onCut(optional) | 剪切的回调 (Trigger when cut) | Function(e: React.MouseEvent \| React.ClipboardEvent) | - |
| onRise(optional) | 移动端检测到软键盘弹起时的回调 (The prefix of class) | Function({ isIOS, isAndroid }) | - |
| onFold(optional) | 移动端检测到软键盘收回时的回调 (The prefix of class) | Function({ isIOS, isAndroid }) | - |
| autoFocus(optional) | 自动聚焦，可以是布尔值，也可传递函数接受参数，而后手动控制聚焦和失焦 (Automatic focus, the value could be a function which will receive the callback that can be control focus or blur by yourself) | boolean \| (handleFocus: (isFocus: boolean) => any) => any | false |

## Example

```javascript
<Input
  type='password'
  className={'hfdc-Input'}
  style={{
    width: '200px'
  }}
  value={val}
  onChange={(e, val) => {
    this.setState({ val });
  }}
  placeholder='I am placeholder'
  autoFocus={handleFocus => {
    if (condition) {
      // focus
      handleFocus(true);
      // blur
      setTimeout(() => handleFocus(false), 2000);
    }
  }}
/>
```

## Tips
使用 `type='password'` 时，为了避免安全问题和React的警告，建议是将组件设置为受控组件，即手动设置 `value` 的值，可参考👆上面的代码片段。
