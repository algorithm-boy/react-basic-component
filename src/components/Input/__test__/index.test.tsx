import * as React from 'react';
import { configure, shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Input from '../index';

configure({ adapter: new Adapter() });

describe('Input - renders', () => {
  it('renders correctly - render', () => {
    const wrapper = render(
      <Input
        placeholder='placeholder'
        defaultValue='defaultValue'
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly - mount', () => {
    const wrapper = mount(
      <div>
        <Input
          placeholder='placeholder'
          value='value'
        />
      </div>
    );
    expect(wrapper.contains(<Input
      placeholder='placeholder'
      value='value'
    />)).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Input - parameters', () => {
  it('parameters check', () => {
    const wrapper = mount(<Input
      type='password'
      id='rc-Input-id'
      prefixCls='rc-Input'
      placeholder='rc-Input-placeholder'
      value='rc-Input-value'
      defaultValue='rc-Input-defaultValue'
      maxLength={500}
      copyFree={false}
      cutFree={false}
      pasteFree={false}
      clipboardFree={false}
    />);

    expect(wrapper.props().type).toEqual('password');
    expect(wrapper.props().id).toEqual('rc-Input-id');
    expect(wrapper.props().prefixCls).toEqual('rc-Input');
    expect(wrapper.props().placeholder).toEqual('rc-Input-placeholder');
    expect(wrapper.props().value).toEqual('rc-Input-value');
    expect(wrapper.props().defaultValue).toEqual('rc-Input-defaultValue');
    expect(wrapper.props().maxLength).toEqual(500);
    expect(wrapper.props().copyFree).toEqual(false);
    expect(wrapper.props().cutFree).toEqual(false);
    expect(wrapper.props().pasteFree).toEqual(false);
    expect(wrapper.props().clipboardFree).toEqual(false);
  });
});

describe('Input - callbacks', () => {
  it('callback check - onChange', done => {
    const wrapper = mount(<Input
      onChange={(e, val) => {
        expect(val).toEqual('test');
        expect(e).toBeTruthy();
        done();
      }}
    />);

    wrapper.simulate('change', { target: { value: 'test' } });
  });

  it('callback check - onKeyDown', done => {
    const wrapper = mount(<Input
      onKeyDown={(e) => {
        expect(e).toBeTruthy();
        expect(e.keyCode).toBe(10);
        done();
      }}
    />);

    wrapper.simulate('keydown', { keyCode: 10 });
  });

  it('callback check - onPressEnter', done => {
    const wrapper = mount(<Input
      onPressEnter={(e) => {
        expect(e).toBeTruthy();
        expect(e.keyCode).toBe(13);
        done();
      }}
    />);

    wrapper.simulate('keydown', { keyCode: 13 });
  });
});
