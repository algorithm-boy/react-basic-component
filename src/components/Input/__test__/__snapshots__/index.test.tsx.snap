// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input - renders renders correctly - mount 1`] = `ReactWrapper {}`;

exports[`Input - renders renders correctly - render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "hfdc-input",
      "maxlength": "Infinity",
      "placeholder": "placeholder",
      "type": "text",
      "value": "defaultValue",
    },
    "children": Array [],
    "name": "input",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "maxlength": undefined,
      "placeholder": undefined,
      "type": undefined,
      "value": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "maxlength": undefined,
      "placeholder": undefined,
      "type": undefined,
      "value": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
